libmuscle/cpp/src/libmuscle/snapshot_manager.cpp

```diff
@@ -0,0 +1,241 @@
+#include <libmuscle/snapshot_manager.hpp>
+
+#include <libmuscle/data.hpp>
+
+#include <algorithm>
+#include <cmath>
+#include <fstream>
+#include <sstream>
+
+// Note: using POSIX for filesystem calls
+// Could be upgraded to std::filesystem when targeting C++17 or later
+#include <unistd.h>
+#include <fcntl.h>
+
+#define MAX_FILE_EXISTS_CHECK 10000
+
+using namespace std::string_literals;
+
+namespace libmuscle { namespace impl {
+
+SnapshotManager::SnapshotManager(
+            ymmsl::Reference const & instance_id,
+            MMPClient & manager,
+            Communicator & communicator,
+            Logger & logger)
+        : instance_id_(instance_id)
+        , manager_(manager)
+        , communicator_(communicator)
+        , logger_(logger)
+        , resume_from_snapshot_()
+        , resume_overlay_()
+        , next_snapshot_num_(1)
+        , snapshot_directory_()
+{
+    // replace identifier[i] by identifier-i to use in snapshot file name
+    // using a dash (-) because that is not allowed in Identifiers
+    safe_id_ = static_cast<std::string>(instance_id_);
+    std::replace(safe_id_.begin(), safe_id_.end(), '[', '-');
+    safe_id_.erase(
+            std::remove(safe_id_.begin(), safe_id_.end(), ']'),
+            safe_id_.end());
+}
+
+namespace {
+
+// Adapted from https://stackoverflow.com/a/2203853
+std::string get_working_path()
+{
+    char temp [ PATH_MAX ];
+
+    if ( getcwd(temp, PATH_MAX) != 0)
+        return std::string ( temp );
+
+    int error = errno;
+    std::ostringstream str;
+    str << "Error retrieving current working directory: " << error;
+    throw std::runtime_error(str.str());
+}
+
+}
+
+Optional<double> SnapshotManager::prepare_resume(
+        Optional<std::string> const & resume_snapshot,
+        Optional<std::string> const & snapshot_directory) {
+    Optional<double> result;
+
+    if (snapshot_directory.is_set() && !snapshot_directory.get().empty()) {
```

Comment:
Fair question. I was mimicking the Python behaviour here where an `Optional[str]` evaluates to false when it is None, or an empty string.

Second part can be removed, because the manager will send either None or a valid Path.